src/hash.cr

```diff
@@ -588,6 +588,61 @@ class Hash(K, V)
     reject! { |key, value| value.nil? }
   end
 
+  # Returns a new hash with all keys converted using the block operation.
+  # The block can change a type of keys.
+  #
+  # ```
+  # hash = {:a => 1, :b => 2, :c => 3}
+  # hash.transform_keys { |key| key.to_s } # => {"A" => 1, "B" => 2, "C" => 3}
+  # ```
+  def transform_keys(&block : K -> K2) forall K2
+    each_with_object({} of K2 => V) do |(key, value), memo|
+      memo[yield(key)] = value
+    end
+  end
+
+  # Returns a new hash with the results of running block once for every value.
+  # The block can change a type of values.
+  #
+  # ```
+  # hash = {:a => 1, :b => 2, :c => 3}
+  # hash.transform_values { |value| value + 1 } # => {:a => 2, :b => 3, :c => 4}
+  def transform_values(&block : V -> V2) forall V2
+    each_with_object({} of K => V2) do |(key, value), memo|
+      memo[key] = yield(value)
+    end
+  end
+
+  # Destructively transforms all keys using a block. Same as transform_keys but modifies in place.
+  # The block cannot change a type of keys.
+  #
+  # ```
+  # hash = {"a" => 1, "b" => 2, "c" => 3}
+  # hash.transform_key! { |key| key.succ }
+  # hash # => {"b" => 1, "c" => 2, "d" => 3}
+  def transform_keys!(&block : K -> K)
```

Comment:
This is breaking the Hash invariant. A call to `rehash` is needed at the end or the final state won't be useful for user.

I don't fully like that the `Entry#key` is changed from `getter` to `property` since that was preventing the invariant to be broken.

I would say to either base the implementation on `rehash` (zeroing the buckets, creating new Entry, iterate the old entries, storing new entries in the right place) or remove this implementation.

If any implementation is added, I would say to assert that `actual.should eq(expected)` and `expected.should eq(actual)` so we can detect if the code messes up with the invariant.